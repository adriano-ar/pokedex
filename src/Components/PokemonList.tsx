import { useEffect, useState } from "react";
import fetchPokemonList from "../Services/fetchPokemonList";
import IPokemon from "../Interfaces/IPokemon";
import Pokemon from "./Pokemon";
import Search from "./Search";
import PokemonDetail from "./PokemonDetail";
import "../Assets/Styles/pokemon-list.scss";

export default function PokemonList() {
    const [pokemonList, setPokemonList] = useState<Array<IPokemon>>([]);
    const [backupPokemonList, setBackupPokemonList] = useState<Array<IPokemon>>(
        []
    );
    const [showingPokemonDetail, setShowingPokemonDetail] = useState(false);
    const [currentPokemon, setCurrentPokemon] = useState(null);
    useEffect(() => {
        getPokemon();
    }, []);

    function getPokemon() {
        fetchPokemonList().then((res) => {
            setPokemonList(res);
            setBackupPokemonList(res);
        });
    }

    function searchPokemon(searchTerm: string) {
        if (searchTerm) {
            setPokemonList(
                backupPokemonList.filter((pokemon) => {
                    return pokemon.name.includes(searchTerm);
                })
            );
        } else {
            setPokemonList(backupPokemonList);
        }
    }

    function showPokemonDetail(id: number) {
        setShowingPokemonDetail(true);
        setCurrentPokemon(id);
    }

    function hidePokemonDetail() {
        setShowingPokemonDetail(false);
        setCurrentPokemon(null);
    }

    return (
        <div id="pokemon-list-wrapper">
            <Search searchPokemon={searchPokemon} />
            <div
                id="pokemon-list"
                className={showingPokemonDetail ? "shrink" : ""}
            >
                <div id="sorter">
                    <div className="stat"></div>
                    <div className="stat">#</div>
                    <div className="stat">Name</div>
                    <div className="stat">Type</div>
                    <div className="stat">HP</div>
                    <div className="stat">Attack</div>
                    <div className="stat">Defense</div>
                    <div className="stat">Sp. Atk</div>
                    <div className="stat">Sp. Def</div>
                    <div className="stat">Speed</div>
                </div>
                {pokemonList.map((pokemon) => (
                    <Pokemon
                        showPokemonDetail={showPokemonDetail}
                        key={pokemon.id}
                        id={pokemon.id}
                        name={pokemon.name}
                        image={pokemon.image}
                        type={pokemon.type}
                        hp={pokemon.hp}
                        attack={pokemon.attack}
                        defense={pokemon.defense}
                        sattack={pokemon.sattack}
                        sdef={pokemon.sdef}
                        speed={pokemon.speed}
                    />
                ))}
            </div>
            <PokemonDetail
                hidePokemonDetail={hidePokemonDetail}
                currentPokemon={currentPokemon}
                showingPokemonDetail={showingPokemonDetail}
            />
        </div>
    );
}
