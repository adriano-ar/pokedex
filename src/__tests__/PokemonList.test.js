/**
 * @jest-environment jsdom
 */
import { expect, test } from "@jest/globals";
import { create } from "react-test-renderer";
import { createRenderer } from "react-test-renderer/shallow";
import PokemonList from "../Components/PokemonList";

const pokemons = [
    {
        name: "bulbasaur",
        id: 1,
        image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
        type: ["grass", "poison"],
        hp: 45,
        attack: 49,
        defense: 49,
        sattack: 65,
        sdef: 65,
        speed: 45,
    },
    {
        name: "charmander",
        id: 4,
        image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/4.png",
        type: ["fire"],
        hp: 39,
        attack: 52,
        defense: 43,
        sattack: 60,
        sdef: 50,
        speed: 65,
    },
    {
        name: "squirtle",
        id: 7,
        image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/7.png",
        type: ["water"],
        hp: 44,
        attack: 48,
        defense: 65,
        sattack: 50,
        sdef: 64,
        speed: 43,
    },
];

test("renders correctly with no pokemons", () => {
    const tree = create(<PokemonList pokemonList={[]} />).toJSON();
    expect(tree).toMatchSnapshot();
});

test("renders correctly with some pokemons", () => {
    const r = createRenderer();
    r.render(<PokemonList pokemonList={pokemons} />);
    expect(r.getRenderOutput()).toMatchSnapshot();
});
