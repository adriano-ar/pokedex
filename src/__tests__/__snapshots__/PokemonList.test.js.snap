// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with no pets 1`] = `
<div
  id="pokemon-list-wrapper"
>
  <div
    id="search-bar"
  >
    <label>
      Name:
    </label>
    <input
      id="search-field"
      onChange={[Function]}
    />
  </div>
  <div
    id="sorter"
  >
    <div
      className="stat"
    />
    <div
      className="stat"
    >
      #
    </div>
    <div
      className="stat"
    >
      Name
    </div>
    <div
      className="stat"
    >
      Type
    </div>
    <div
      className="stat"
    >
      HP
    </div>
    <div
      className="stat"
    >
      Attack
    </div>
    <div
      className="stat"
    >
      Defense
    </div>
    <div
      className="stat"
    >
      Sp. Atk
    </div>
    <div
      className="stat"
    >
      Sp. Def
    </div>
    <div
      className="stat"
    >
      Speed
    </div>
  </div>
  <div
    className=""
    id="pokemon-list"
  />
  <div
    className=""
    data-testid="pokemonDetailView"
    id="pokemon-detail"
  >
    <div
      data-testid="closeButton"
      id="close-button"
      onClick={[Function]}
    />
    <div
      id="pokemon-name"
    >
      
    </div>
    <img
      id="pokemon-pic"
      src=""
    />
    <div
      className="stats-title"
    >
      Pokédex data
    </div>
    <div
      className="stat-row"
    >
      <div
        className="stat-label"
      >
        National № 
      </div>
      <div
        className="stat"
      >
        000
      </div>
    </div>
    <div
      className="stat-row"
    >
      <div
        className="stat-label"
      >
        Type
      </div>
      <div
        className="stat"
      />
    </div>
    <div
      className="stat-row"
    >
      <div
        className="stat-label"
      >
        Height 
      </div>
      <div
        className="stat"
      >
        0 m
      </div>
    </div>
    <div
      className="stat-row"
    >
      <div
        className="stat-label"
      >
        Weight 
      </div>
      <div
        className="stat"
      >
        0 kg
      </div>
    </div>
    <div
      className="stat-row"
    >
      <div
        className="stat-label"
      >
        Abilities 
      </div>
      <div
        className="stat"
      />
    </div>
    <div
      className="stats-title"
    >
      <div
        className="stat-label"
      >
        Base stats
      </div>
    </div>
    <div
      className="stat-row"
    >
      <div
        className="stat-label"
      >
        HP 
      </div>
      <div
        className="stat"
      >
        0
      </div>
      <div
        className="stat-bar"
      >
        <div
          className="stat-bar-value"
          style={
            {
              "width": "0%",
            }
          }
        />
      </div>
    </div>
    <div
      className="stat-row"
    >
      <div
        className="stat-label"
      >
        Attack 
      </div>
      <div
        className="stat"
      >
        0
      </div>
      <div
        className="stat-bar"
      >
        <div
          className="stat-bar-value"
          style={
            {
              "width": "0%",
            }
          }
        />
      </div>
    </div>
    <div
      className="stat-row"
    >
      <div
        className="stat-label"
      >
        Defense
      </div>
       
      <div
        className="stat"
      >
        0
      </div>
      <div
        className="stat-bar"
      >
        <div
          className="stat-bar-value"
          style={
            {
              "width": "0%",
            }
          }
        />
      </div>
    </div>
    <div
      className="stat-row"
    >
      <div
        className="stat-label"
      >
        Sp. Attack
      </div>
      <div
        className="stat"
      >
        0
      </div>
      <div
        className="stat-bar"
      >
        <div
          className="stat-bar-value"
          style={
            {
              "width": "0%",
            }
          }
        />
      </div>
    </div>
    <div
      className="stat-row"
    >
      <div
        className="stat-label"
      >
        Sp. Def 
      </div>
      <div
        className="stat"
      >
        0
      </div>
      <div
        className="stat-bar"
      >
        <div
          className="stat-bar-value"
          style={
            {
              "width": "0%",
            }
          }
        />
      </div>
    </div>
    <div
      className="stat-row"
    >
      <div
        className="stat-label"
      >
        Speed 
      </div>
      <div
        className="stat"
      >
        0
      </div>
      <div
        className="stat-bar"
      >
        <div
          className="stat-bar-value"
          style={
            {
              "width": "0%",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly with some pets 1`] = `
<div
  id="pokemon-list-wrapper"
>
  <Search
    searchPokemon={[Function]}
  />
  <div
    id="sorter"
  >
    <div
      className="stat"
    />
    <div
      className="stat"
    >
      #
    </div>
    <div
      className="stat"
    >
      Name
    </div>
    <div
      className="stat"
    >
      Type
    </div>
    <div
      className="stat"
    >
      HP
    </div>
    <div
      className="stat"
    >
      Attack
    </div>
    <div
      className="stat"
    >
      Defense
    </div>
    <div
      className="stat"
    >
      Sp. Atk
    </div>
    <div
      className="stat"
    >
      Sp. Def
    </div>
    <div
      className="stat"
    >
      Speed
    </div>
  </div>
  <div
    className=""
    id="pokemon-list"
  />
  <PokemonDetail
    currentPokemon={null}
    hidePokemonDetail={[Function]}
    showingPokemonDetail={false}
  />
</div>
`;
